### DISCUSSION TITLE
Git push from a manual workflow it doesn't trigger other on-push workflow

### DISCUSSION THREAD

**Discussion by haizaar** [community/community#25424] [Actions]
<p>Good day,</p>
<p>I have workflow in my repo to trigger build on push (including tags). When I create and push tags from my laptop it’s triggered just fine.<br>
I then went ahead and created a manual (<code>workflow_dispatch</code>) workflow to tag a new release in my repo (similar to what I do on my laptop). However when I run this workflow and it does tag/push successfully the on-push workflow <em>doesn’t trigger</em>.</p>
<p>What am I missing?</p>
<p>My manual workflow is below. Using the same <code>make bumpver</code> from my laptop shell works as expected triggering build in GH.</p>
<pre><code class="lang-auto">name: Tag

on:
  workflow_dispatch: {}

jobs:
  tag:
    name: Tag new version
    runs-on: ubuntu-latest

    steps:
    - name: Checkout
      uses: actions/checkout@v2
      with:
        fetch-depth: 0

    - name: Tag
      run: |
        set -x
        git config --local user.email "action@github.com"
        git config --local user.name "${{ github.actor }}"
        export PATH=$HOME/.local/bin:$PATH
        python3 -m pip install --upgrade pip --user
        which pip
        pip install bumpver --user
        make release
      env:
        LC_ALL: C.UTF-8
</code></pre>

**Comment by airtower-luna** [comment]
<p>The documentation on <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/authentication-in-a-workflow#using-the-github_token-in-a-workflow">Using the <code>GITHUB_TOKEN</code> in a workflow</a> explains this behavior:</p>
<blockquote>
<p>When you use the repository’s <code>GITHUB_TOKEN</code> to perform tasks on behalf of the GitHub Actions app, events triggered by the <code>GITHUB_TOKEN</code> will not create a new workflow run. This prevents you from accidentally creating recursive workflow runs. For example, if a workflow run pushes code using the repository’s <code>GITHUB_TOKEN</code> , a new workflow will not run even when the repository contains a workflow configured to run when <code>push</code> events occur.</p>
</blockquote>
<p>It you want the push to trigger a new workflow you need to <a href="https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/creating-a-personal-access-token">use a PAT (personal access token)</a> for authentication instead. Or you could try adding a <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#workflow_run"><code>workflow_run</code> trigger</a> to the build workflow to run it when the tag workflow completes.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a>,</p>
<p>By default, the <a href="https://github.com/actions/checkout" rel="noopener nofollow ugc">checkout</a> action uses <strong>${{ github.token }}</strong> (that is the <strong>GITHUB_TOKEN</strong>) as the authentication token, and set <strong>persist-credentials</strong> as <strong>true</strong>.<br>
In this situation, the <strong>GITHUB_TOKEN</strong> will be used to configure the <strong>local git config</strong>, and the git commands in the subsequent steps of the same job will use this <strong>GITHUB_TOKEN</strong> as the authentication token. Even if you provide another token (such as a PAT) in the subsequent steps.<br>
More details, see the input parameters of checkout action:</p>
<ul>
<li><a href="https://github.com/actions/checkout/blob/c952173edf28a2bd22e1a4926590c1ac39630461/action.yml#L12-L24" rel="noopener nofollow ugc"><strong>token</strong></a></li>
<li><a href="https://github.com/actions/checkout/blob/c952173edf28a2bd22e1a4926590c1ac39630461/action.yml#L48-L50" rel="noopener nofollow ugc"><strong>persist-credentials</strong></a></li>
</ul>
<p>As <a class="mention" href="/airtower-luna">@airtower-luna</a> mentioned, the <strong>GITHUB_TOKEN</strong> can’t trigger new workflow run, you need to use a personal access token (PAT).</p>
<p>In your case, if you want to trigger the workflow with the push, you can try like as this:<br>
Use a PAT as the value of the input <strong>token</strong> on the checkout action in your workflow, and <strong>persist-credentials</strong> still is <strong>true</strong>. In this way the PAT will be used to configure the <strong>local git config</strong>.</p>
<pre><code class="lang-yaml">steps:
  - name: Checkout
    uses: actions/checkout@v2
    with:
      token: ${{ secrets.GITHUB_PAT }}
      fetch-depth: 0

- name: Tag
  run: |
    . . .
</code></pre>

**Comment by haizaar** [comment]
<p>Thank you for the quick and educating responses. I chose to use <code>workflow_run</code> trigger as suggested and it did the job beautifully.</p>

**Comment by haizaar** [comment]
<p>Unfortunately my conclusions were premature:</p>
<ul>
<li>
<code>workflow_run</code> trigger indeed triggers the build workflow, but it runs on HEAD, not on the created tag which misses the point obviously (my build workflow has “release” as its final step that runs only if the build git ref is tag)</li>
<li>Using PAT is a security breach IMHO - if I generate a PAT and load it as GH secret for GH Actions to use, it gives anyone with write access to the repo access to this token and hence to everything this token has a access to (including my other repos, etc.). I wish there could be a way to scope a PAT to a particular repo only but I can’t find any sane [1] way to do it</li>
</ul>
<p>So I’m back to square one :expressionless:</p>
<p>[1] <a href="https://github.community/t/limiting-scope-of-a-pat-to-a-single-repository/3129/2" class="inline-onebox">Limiting scope of a PAT to a single repository</a></p>

**Comment by haizaar** [comment, marked as answer]
<p>After more research, using a deploy key results in quite reasonable setup to achieve what I want:</p>
<ul>
<li>Created dedicated SSH key pair on my computer.</li>
<li>Registered the public part as a deploy key for the repo in question <em>with write access</em>
</li>
<li>Loaded the private part as the <code>DEPLOY_KEY_CI_TAGGING</code> GH Secret</li>
</ul>
<p>Here is the full workflow:</p>
<pre><code class="lang-auto">name: Tag

on:
  workflow_dispatch: {}

jobs:
  tag:
    name: Tag new version
    runs-on: ubuntu-latest

    steps:
    - name: Checkout
      uses: actions/checkout@v2
      with:
        fetch-depth: 0

    # https://github.community/t/git-push-from-a-manual-workflow-it-doesnt-trigger-other-on-push-workflow/139190/5
    - name: Switch git credentials to RW
      uses: webfactory/ssh-agent@v0.4.1
      with:
        ssh-private-key: ${{ secrets.DEPLOY_KEY_CI_TAGGING }}

    - name: Tag
      run: |
        set -x
        git config --local user.email "action@github.com"
        git config --local user.name "${{ github.actor }}"
        # Force using the SSH credentials from the previous step
        git remote set-url origin git@github.com:${{ github.repository }}.git
        make release  # create tag, push tag, etc.
      env:
        LC_ALL: C.UTF-8
</code></pre>
<p>Thank you all again for the information provided.</p>
